src/main/java/org/broadinstitute/hellbender/tools/copynumber/CollectFragmentCounts.java

```diff
@@ -0,0 +1,223 @@
+ * how many different fragment centers intersect with this interval. The start and end positions of fragments are
+ * inferred from read information. We only allow properly paired, first of pair reads - thus we do not double count
+ * and we exclude reads whose fragment's position cannot be automatically inferred from its SAM record.
+ *
+ * @author Andrey Smirnov &lt;asmirnov@broadinstitute.org&gt;
+ */
+@CommandLineProgramProperties(
+        summary = "Collect fragment counts, by counting how many fragment centers intersect with a" +
+                " given interval. The fragments are inferred from SAM records of only properly paired intervals.",
+        oneLineSummary = "Collect fragment counts",
+        programGroup = CopyNumberProgramGroup.class
+)
+public class CollectFragmentCounts extends ReadWalker {
+
+    private static Logger logger = LogManager.getLogger(CollectFragmentCounts.class);
+    private SAMSequenceDictionary sequenceDictionary = getBestAvailableSequenceDictionary();
+
+    public static final String COLUMN_SEPARATOR = "\t";
+    public static final String LINE_SEPARATOR = "\n";
+
+    @Argument(
+            doc = "Output fragment-counts file",
+            fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME,
+            shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME
+    )
+    private File outputCountsFile = null;
+
+    /**
+     * Per interval counts, ordered according to {@link CollectFragmentCounts#cachedIntervalList}
+     */
+    private int[] intervalCounts;
+
+    /**
+     * List of non-overlapping searchable intervals
+     */
+    private CachedBinarySearchIntervalList<SimpleInterval> cachedIntervalList;
+
+    /**
+     * Name of the sample contained in the BAM file
+     */
+    private String sampleName;
+
+    @Override
+    public List<ReadFilter> getDefaultReadFilters() {
+        final List<ReadFilter> filters = new ArrayList<>(super.getDefaultReadFilters());
+        filters.add(ReadFilterLibrary.MAPPED);
+        filters.add(ReadFilterLibrary.NON_ZERO_REFERENCE_LENGTH_ALIGNMENT);
+        filters.add(ReadFilterLibrary.NOT_DUPLICATE);
+        filters.add(ReadFilterLibrary.FIRST_OF_PAIR); // this will make sure we don't double count
+        filters.add(ReadFilterLibrary.PROPERLY_PAIRED);
+        // this will only keep reads in pairs that are properly oriented and mapped on same chromosome
+        // and lie within a few standard deviations from the mean of fragment size distributions
+        return filters;
+    }
+
+    @Override
+    public boolean requiresIntervals() {
+        return true;
+    }
+
+    @Override
+    public void onTraversalStart() {
+        if (readArguments.getReadFilesNames().size() != 1) {
+            throw new UserException("This tool only accepts a single bam/sam/cram as input");
+        }
+
+        final SampleCollection sampleCollection = new SampleCollection(getHeaderForReads());
+        if (sampleCollection.sampleCount() > 1) {
```

Comment:
There is some sample-name code in #3716 that we can use for things like this.

Reply:
However, note that even that code is a bit of a stopgap---we should eventually do this right with full BAM-style headers.